tensor2tensor/data_generators/librispeech.py

```diff
@@ -69,41 +89,233 @@ def _get_audio_data(filepath):
+    i = 0
+    for url, subdir in datasets:
+      filename = os.path.basename(url)
+      compressed_file = generator_utils.maybe_download(tmp_dir, filename, url)
+
+      read_type = "r:gz" if filename.endswith("tgz") else "r"
+      with tarfile.open(compressed_file, read_type) as corpus_tar:
+        # Create a subset of files that don't already exist.
+        #   tarfile.extractall errors when encountering an existing file
+        #   and tarfile.extract is extremely slow
+        members = []
+        for f in corpus_tar:
+          if not os.path.isfile(os.path.join(tmp_dir, f.name)):
+            members.append(f)
+        corpus_tar.extractall(tmp_dir, members=members)
   
-    data_dir = os.path.join(tmp_dir, "LibriSpeech", subdir)
-    data_files = _collect_data(data_dir, "flac", "txt")
-    data_pairs = data_files.values()
-    for media_file, text_data in sorted(data_pairs)[start_from:]:
-      if how_many > 0 and i == how_many:
-        return
-      i += 1
-      audio_data, sample_count, sample_width, num_channels = _get_audio_data(
-          media_file)
-      label = [ord(c) for c in text_data] + eos_list
-      yield {
-          "inputs": audio_data,
-          "audio/channel_count": [num_channels],
-          "audio/sample_count": [sample_count],
-          "audio/sample_width": [sample_width],
-          "targets": label
-      }
\ No newline at end of file
+      data_dir = os.path.join(tmp_dir, "LibriSpeech", subdir)
+      data_files = _collect_data(data_dir, "flac", "txt")
+      data_pairs = data_files.values()
+      for media_file, text_data in sorted(data_pairs)[start_from:]:
+        if how_many > 0 and i == how_many:
+          return
+        i += 1
+        audio_data, sample_count, sample_width, num_channels = _get_audio_data(
+            media_file)
+        label = [ord(c) for c in text_data] + eos_list
+        yield {
+            "inputs": audio_data,
+            "audio/channel_count": [num_channels],
+            "audio/sample_count": [sample_count],
+            "audio/sample_width": [sample_width],
+            "targets": label
+        }
+        
+        
+  def generate_data(self, data_dir, tmp_dir, task_id=-1):
+    train_paths = self.training_filepaths(data_dir, self.num_shards, shuffled=False)
+    dev_paths = self.dev_filepaths(data_dir, self.num_dev_shards, shuffled=False)
+    if self.use_train_shards_for_dev:
+      all_paths = train_paths + dev_paths
+      generator_utils.generate_files(self.generator(data_dir, tmp_dir, True), all_paths)
+      generator_utils.shuffle_dataset(all_paths)
+    else:
+      generator_utils.generate_dataset_and_shuffle(
+          self.generator(data_dir, tmp_dir, True), train_paths,
+          self.generator(data_dir, tmp_dir, False), dev_paths)
+      
+    
+  def hparams(self, defaults, unused_model_hparams):
+    p = defaults    
+    p.stop_at_eos = int(False)
+    p.input_modality = { "inputs": ("audio:librispeech_modality", None) }      

```

Comment:
registry.Modalities.AUDIO

Reply:
Wouldn't that result in the base Audio modality being used, and bypass the custom signal processing added to `bottom()`?